pkg/syncer/broker/federator.go

```diff
@@ -0,0 +1,166 @@
+package broker
+
+import (
+	"fmt"
+	"reflect"
+
+	"github.com/pkg/errors"
+	"github.com/submariner-io/admiral/pkg/federate"
+	"github.com/submariner-io/admiral/pkg/log"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/util/retry"
+	"k8s.io/klog"
+)
+
+const (
+	metadataField = "metadata"
+	labelsField   = "labels"
+)
+
+type federator struct {
+	dynClient       dynamic.Interface
+	restMapper      meta.RESTMapper
+	brokerNamespace string
+	localClusterID  string
+}
+
+var keepMetadataFields = map[string]bool{"name": true, "namespace": true, labelsField: true, "annotations": true}
+
+func NewFederator(localClusterID string) (federate.Federator, error) {
+	restConfig, namespace, err := buildConfig()
+	if err != nil {
+		return nil, err
+	}
+
+	dynClient, err := dynamic.NewForConfig(restConfig)
+	if err != nil {
+		return nil, fmt.Errorf("error creating dynamic client: %v", err)
+	}
+
+	restMapper, err := buildRestMapper(restConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	return &federator{
+		dynClient:       dynClient,
+		restMapper:      restMapper,
+		brokerNamespace: namespace,
+		localClusterID:  localClusterID,
+	}, nil
+}
+
+// Use for testing purposes only
+func NewTestFederator(dynClient dynamic.Interface, restMapper meta.RESTMapper, brokerNamespace string, localClusterID string) federate.Federator {
+	return &federator{
+		dynClient:       dynClient,
+		restMapper:      restMapper,
+		brokerNamespace: brokerNamespace,
+		localClusterID:  localClusterID,
+	}
+}
+
+func (f *federator) Distribute(resource runtime.Object) error {
+	toDistribute, resourceClient, err := f.toUnstructured(resource)
+	if err != nil {
+		return err
+	}
+
+	if f.localClusterID != "" {
+		setNestedField(toDistribute.Object, f.localClusterID, metadataField, labelsField, federate.ClusterIDLabelKey)
+	}
+
+	return retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		existing, err := resourceClient.Get(toDistribute.GetName(), metav1.GetOptions{})
+		if apierrors.IsNotFound(err) {
+			removeFieldsForCreate(toDistribute)
+			toDistribute.SetNamespace(f.brokerNamespace)
+
+			klog.V(log.DEBUG).Infof("Creating resource: %#v", toDistribute)
+
+			_, err := resourceClient.Create(toDistribute, metav1.CreateOptions{})
+			if apierrors.IsAlreadyExists(err) {
+				klog.V(log.DEBUG).Infof("Resource %q already exists", toDistribute.GetName())
+				return nil
```

Comment:
Don’t we want to update it here? (I mean, fall through to the update code.)

Reply:
I thought of that but this means it was created out-of-band so unsure whether we should automatically replace it or if the caller would want to merge in this case. Perhaps we should just return the error and let the caller decide how to handle it. If they want to replace then they can retry the call. In our uses cases this `shouldn't` happen.

Reply:
Thinking about this again, I think we should just update here. The caller expects its version of the resource to be applied if it succeeds and I think returning `AlreadyExists` would be confusing. This is just a timing edge case that is really equivalent to the preceding `Get` returning success a split second earlier if the resource was created out-of-band.

Reply:
So I'll return a conflict error to cause a retry.

Reply:
Done